### ISSUE TITLE
event notifications in the shell

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#11437]
I've started to think about a design for notifications in the shell and I think we need the following in a notifications API that enables pages to talk to the shell and the shell to talk to pages:

## Communication from page to shell

- **Notifications** from a page
  - Includes (items are optional unless otherwise specified):
    - **Title** _(mandatory)_
    - **String** additional details
    - **Details** longer string, usually used for system error messages, not shown by default
    - **Type** of notification (info, warning, error) _(defaults to info if not specified)_
    - **Actions** in the notifications (primary & secondary, such as "Apply and Restart", "View Details", etc.) — this should be optional, as not all notifications will have actions
    - **Persistence** — if there's an ongoing error, like disk space problems, it shouldn't be dismissable
    - **Origin** page (& subpage, if applicable) _(mandatory)_
    - **Highlight** selector (probably a `querySelectorAll()` compatible string, most likely an ID): used when switching to the origin page to show a place where an action needs to take place; could be handled by the receiving page
    - **Display** area (notification tray, nav, both, neither — _defaults to tray + nav_)
      - some notifications would only show up in the tray (without a menu status dot)
      - some would only show up as a dot: if there's output from a long-running terminal, you don't want a popup, but having something like a blue dot next to the terminal entry on the sidebar would be nice (if you're in another tab)
      - some could be background notifications, so they skip the tray and nav dot, but are available for the overview page to skim and present — basically, a ghost notification in the shell :ghost::exclamation::arrow_right::shell:

- **Badges** in the sidebar
  - For the host menu especially, but also for top-level apps like Dashboard and Composer
  - This would probably piggyback on the general notification API, except it would display a status dot with the highest level of error state (and a hover would probably show notification titles)

- A way to **clear notifications**
  - **Clear individual** notification
  - **Clear all** notifications
  - **Clear navigation dot** when navigating to a page

- Pages should also be able to **query the shell** for notifications, so one page is aware of what's happening in another — this would be useful for showing health-related issues in the overview page

## Communication from shell to page

These are hinted at above. In order to take actions, we need to talk to the page too.

- Ability to **query a page** for information
  - Examples: Ask Software Updates if there are updates, Ask storage if there are any issues (full disk, offline disks, etc.)
    - This could be done in register listener type of a style, or it could directly dispatch an action from another page (it's up to the implementation)

- **Perform actions** from notifications
  - This can directly start a process from the notification
  - An action may also change the active page, like a "View more" action
  - (An action can also both start a process and change the current page)

**Comment by garrett** [comment]
Thoughts about the visual design:

- I'm currently thinking that we want to have notifications pop up above the shell, next to the notification area. After a short delay, they disappear (unless the mouse is hovering over the notification, then the timeout is reset and started again when the mouse moves away), but an animation would show that they are available in the notification tray.

- Notification dots (or "blips" perhaps) can also show up in the navigation. They're cleared when the notification is cleared (either removed or resolved) and/or someone navigates to the page.

Yeah, it's pretty much that simple at a high-level, but we need most of the above to be able to do it properly.

**Comment by andreasn** [comment]
Some global notifications from other products using Patternfly:
* [Cloudforms](http://talk.manageiq.org/t/ux-feedback-notification-drawer/2928)
* [Gluster](https://access.redhat.com/documentation/en-us/red_hat_gluster_storage/3.4/html/monitoring_guide/alerts_and_notifications)
* [Foreman/Satellite](https://projects.theforeman.org/attachments/download/2592/notifications.png)

[Patternfly design guidelines](https://www.patternfly.org/pattern-library/communication/notification-drawer/#design)

**Comment by andreasn** [comment]
Do you think it makes sense to include long-running-operations in this piece of work, or is exploding things in scope too early?
I'm thinking specifically about things like restoring a RAID or building an image in Composer.

**Comment by garrett** [comment]
@andreasn: I think it makes sense to start thinking about how that would work, even if they might not be implemented in phase 1.

For long operations, we might want to think about having them show up outside of the notifications area (perhaps to the left).

Basically, like:
> Restoring RAID [#####----]

_next_ to the notification area in the top bar.

Perhaps we'd implement it as a notification with the following attributes:

- Title: Restoring RAID
- Persist: true
- Origin: Storage
- Highlight: #ct-storage-raid
- **Progress**: 65%
- Display: **Topbar**

(Bolded parts that aren't in the above proposed spec at the moment)

Progress could be set as -1 for a spinner, otherwise 0 - 100 would be done. Omitted progress (or set to null) would be a topbar notification without progress.

**Comment by martinpitt** [comment]
@garrett: The API/mechanics has landed, plus two consumers (Services and Software Updates pages), plus Insights integration in RHEL. Do you still want to keep this open for some details, or open new issues for specific kinds of notifications?

**Comment by mvollmer** [comment]
We only have state notifications, but no event ones yet.  (state = "there is something wrong here, come have a look and try to make it go away", event = "this just happened, do you care?"  We need new API for those.

**Comment by martinpitt** [comment]
@garrett , @mvollmer : Do you still actually care about this? IMHO this is covered well enough with the current state API -- e.g. the state could change to "Restoring RAID" and further to "RAID restored", and if people care about the details/progress, they can look at the storage page. I'm inclined to close this.

**Comment by martinpitt** [comment]
Please yell and reopen if you still care about this, but so far we didn't run into a use case for that. Thanks!

**Comment by garrett** [comment]
**Summary**: Closing this is good. :+1: 

---

**Detailed answer** (in case people reading this wonder):

While implementing, we decided to not do notification management. 

We show the current state of issues in the health card (instead of notifications, which are transient). We also show an indicator in the navigation (as mentioned in this issue).

I remember that some people have asked for browser-based notifications, which could be an enhancement, and could include things that aren't warnings, for new events that happen or long running tasks (VM installation when the page isn't focused, disk formatting, etc.) But this issue did not cover browser-based notifications.

I think we did implement what we wanted to do and this issue has been addressed, even if a little different from the original spec... and that's fine and intentional in this case.
